src/jit/lower.cpp

```diff
@@ -338,24 +338,96 @@ void Lowering::DecomposeNode(GenTreePtr* pTree, Compiler::fgWalkData* data)
+                    loResult->AsCast()->gtCastType = TYP_UINT;
+                    loResult->gtType = TYP_INT;
+
+                    hiResult = comp->gtNewIconNode(0);
+                }
+                else
                 {
+                    comp->fgInsertEmbeddedFormTemp(&tree->gtOp.gtOp1);
+
                     loResult = tree->gtGetOp1();
-                    hiResult = new (comp, GT_CNS_INT) GenTreeIntCon(TYP_INT, 0);
+
+                    GenTree* loCopy = comp->gtNewLclvNode(loResult->AsLclVarCommon()->GetLclNum(), TYP_INT);
+                    GenTree* shiftBy = comp->gtNewIconNode(31);
+                    hiResult = comp->gtNewOperNode(GT_RSH, TYP_INT, loCopy, shiftBy);
+                    comp->gtSetEvalOrder(hiResult);
+                    comp->fgSetTreeSeq(hiResult);
+                    comp->lvaIncRefCnts(loCopy);
+
                     comp->fgSnipNode(curStmt, tree);
                 }
+            }
+            else if (varTypeIsLong(srcType))
+            {
+                if (tree->gtOverflow() && (varTypeIsUnsigned(srcType) != varTypeIsUnsigned(dstType)))
+                {
+                    //
+                    // When casting between long types an overflow check is needed only if the types
+                    // have different signedness. In both cases (long->ulong and ulong->long) we only
+                    // need to check if the high part is negative or not. Use the existing cast node
+                    // to perform a int->uint cast of the high part to take advantage of the overflow 
+                    // check provided by codegen.
+                    //
+
+                    noway_assert(srcOp->OperGet() == GT_LONG);
+
+                    loResult = srcOp->gtGetOp1();
+
+                    hiResult = tree;
+                    hiResult->gtType = TYP_INT;
+                    hiResult->AsCast()->gtCastType = TYP_UINT;
+                    hiResult->gtFlags &= ~GTF_UNSIGNED;
+                    hiResult->gtOp.gtOp1 = srcOp->gtGetOp2();
+
+                    comp->fgSnipNode(curStmt, srcOp);
+                    newTree = new (comp, GT_LONG) GenTreeOp(GT_LONG, TYP_LONG, loResult, hiResult);
+                    SimpleLinkNodeAfter(hiResult, newTree);
+
+                    //
+                    // Change loResult and hiResult to null so the code after the switch doesn't 
+                    // attempt to insert hiResult after loResult. We only need it to insert newTree.
+                    //
+
+                    loResult = nullptr;
+                    hiResult = nullptr;
+                }
                 else
                 {
-                    NYI("Lowering of signed cast TYP_INT->TYP_LONG");
+                    NYI("Unimplemented long->long no-op cast");
                 }
-                break;
-            default:
+            }
+            else
+            {
                 NYI("Unimplemented type for Lowering of cast to TYP_LONG");
-                break;
```

Comment:
The main remaining types are float and double, such casts should be implemented by helper calls so it's likely that this won't be hit either.